io/jvm/src/main/scala/fs2/io/ioplatform.scala

```diff
@@ -26,12 +26,49 @@ import cats.effect.kernel.{Async, Outcome, Resource, Sync}
 import cats.effect.kernel.implicits._
 import cats.effect.kernel.Deferred
 import cats.syntax.all._
+import fs2.internal.ThreadFactories
 import fs2.io.internal.PipedStreamBuffer
 
 import java.io.{InputStream, OutputStream}
+import java.util.concurrent.Executors
 
 private[fs2] trait ioplatform extends iojvmnative {
 
+  private[this] lazy val stdinExecutor =
+    Executors.newSingleThreadExecutor(ThreadFactories.named("fs2-stdin", true))
+
+  /** Stream of bytes read asynchronously from standard input. */
+  def stdin[F[_]](bufSize: Int)(implicit F: Async[F]): Stream[F, Byte] =
+    readInputStreamGeneric(
+      F.pure(System.in),
+      F.delay(new Array[Byte](bufSize)),
+      false
+    ) { (is, buf) =>
+      F.async[Int] { cb =>
+        F.delay {
+          val task: Runnable = () => cb(Right(is.read(buf)))
+          stdinExecutor.submit(task)
+        }.map { fut =>
+          Some(F.delay {
+            fut.cancel(false)
+            ()
+          })
+        }
+      }
+    }
```

Comment:
Ok, I think this works :) we run all reads from stdin on a custom, single-threaded executor. That way at most one runs at a time, so we never leak threads. We also can't leak fibers, since all tasks (except the one that's running) can always be canceled and removed.